pytest_operator/plugin.py

```diff
@@ -347,46 +354,158 @@ def download_resource(self, resource, destination: Path):
+    """Raise when referencing to a model that doesn't exist."""
+
+
+class ModelInUseError(Exception):
+    """Raise when trying to add a model alias which already exists."""
+
+
+@dataclasses.dataclass
+class ModelState:
+    model: Model
+    keep: bool
+    controller_name: str
+    cloud_name: Optional[str]
+    model_name: str
+    config: dict = None
+    tmp_path: Optional[Path] = None
+
+    @property
+    def full_name(self) -> str:
+        return f"{self.controller_name}:{self.model_name}"
+
+
 class OpsTest:
     """Utility class for testing Operator Charms."""
 
-    _instance = None  # store instance so we can tell if it's been used yet
+    _instance = None  # store instance, so we can tell if it's been used yet
 
     def __init__(self, request, tmp_path_factory):
         self.request = request
-        self.tmp_path: Path = tmp_path_factory.mktemp(self.default_model_name)
-        log.info(f"Using tmp_path: {self.tmp_path}")
+        self._tmp_path_factory = tmp_path_factory
+        self._global_tmp_path = None
 
         # Flag indicating whether all subsequent tests should be aborted.
         self.aborted = False
 
         # Flag for using destructive mode or not for charm builds.
         self.destructive_mode = request.config.option.destructive_mode
 
+        # Config options to determine first model specs used by tests
+        self._orig_model_alias = request.config.option.model_alias
+        self._init_cloud_name = request.config.option.cloud
+        self._init_model_name = request.config.option.model
+        self._init_keep_model = request.config.option.keep_models
+
         # These may be modified by _setup_model
-        self.cloud_name = request.config.option.cloud
         self.controller_name = request.config.option.controller
-        self.model_name = request.config.option.model
-        self.keep_model = request.config.option.keep_models
-        self.model_config = request.config.option.model_config
+        self._init_model_config = request.config.option.model_config
 
         # Flag for enabling the juju-crashdump
         self.crash_dump = not request.config.option.no_crash_dump
         self.crash_dump_output = request.config.option.crash_dump_output
 
         # These will be set by _setup_model
-        self.model_full_name = None
-        self.model = None
         self.jujudata = None
-        self._controller = None
+        self._controller: Optional[Controller] = None
+
+        # maintains a set of all models connected by this fixture
+        self._current_alias = None
+        self._models: MutableMapping[str, ModelState] = {}
+
+    @contextlib.contextmanager
+    def model_context(self, alias: str) -> Generator[Model, None, None]:
+        """
+        Analog to `juju switch` where the focus of the current model is moved.
+        """
+        prior = self.current_alias
+        model = self._switch(alias)
+        yield model
+        self._switch(prior)
```

Comment:
in order to use an existing model, use this as a context generator

imagine there's a controller with 2 clouds "my-machine-cloud" and under it, a "my-k8s-cloud"

using the model aliases `main` and `k8s-model`, you can test the attributes of the models

```python
with ops_test.model_context("main") as model:
    assert "kubernetes-worker" in model.applications
    assert ops_test.cloud_name == "my-machine-cloud"
with ops_test.model_context("k8s-model") as model:
    assert "hello-kubecon" in model.applications
    assert ops_test.cloud_name == "my-k8s-cloud"
```